codepropertygraph/src/main/scala/io/shiftleft/passes/DiffGraphProtoSerializer.scala

```diff
@@ -136,11 +124,21 @@ class DiffGraphProtoSerializer {
       .setPropertiesHash(ByteString.copyFrom(DiffGraph.propertiesHash(edge.asInstanceOf[OdbEdge])))
       .build
 
-  private def removeNodePropertyProto(nodeId: Long, propertyKey: String) =
-    DiffGraphProto.RemoveNodeProperty.newBuilder
-      .setKey(nodeId)
-      .setName(NodePropertyName.valueOf(propertyKey))
-      .build
+
+  private def removeNodePropertyProto(nodeId: Long, propertyKey: String) = {
+    if (propertyKey(0).isLower) {
```

Comment:
This disguishment looks fishy. Having a property start with lower case to encode it as CONTAINED_REF is a way to implicit encoding.

Reply:
It _is_ super fishy!

The correct approach would be to check whether the property exists in the enum. However, afaiu proto doesn't have an api for that (unless we want to use exceptions for regular control flow, but that would be even uglier)

The even better thing would be to have "real" proto properties -- but that would require us to remodel everything within the same mindset of "cpg-style" properties (i.e. properties exist and are documented independently of the nodes that carry them -- as opposed to "class-style" properties, where it is conceptually nonsensical to even think about two classes with the same property, unless the property is declared in a common ancestor in the inheritance hierarchy). If I start remodeling the object hierarchy, then I'll vanish for weeks and rewrite too much code (and @fabsx00 would rightfully crucify me for exhibiting not-invented-by-myself syndrome), so I tried to stick to rewriting only code that is either both objectively awful and a bad offender in profiling, or needs to be rewritten to be compatible with the changes.

Do you have other better ideas? Did I just fail to rtfm java protobuf?

Reply:
> The correct approach would be to check whether the property exists in the enum. However, afaiu proto doesn't have an api for that

How about something like `io.shiftleft.proto.cpg.Cpg.NodePropertyName.values.toSet` (and then hold onto it somewhere as long as you need it)?